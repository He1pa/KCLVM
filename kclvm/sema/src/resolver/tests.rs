use super::Options;
use super::Resolver;
use crate::builtin::BUILTIN_FUNCTION_NAMES;
use crate::pre_process::pre_process_program;
use crate::resolver::lint;
use crate::resolver::resolve_program;
use crate::resolver::scope::*;
use crate::ty::Type;
use indexmap::IndexSet;
use kclvm_ast::ast;
use kclvm_error::*;
use kclvm_parser::{load_program, parse_program};
use std::rc::Rc;

#[test]
fn test_scope() {
    let mut scope = builtin_scope();
    for name in BUILTIN_FUNCTION_NAMES {
        let obj = scope.lookup(name).unwrap();
        let obj_ref = obj.borrow_mut();
        assert!(obj_ref.ty.is_func());
    }
    for name in BUILTIN_FUNCTION_NAMES {
        scope.set_ty(name, Rc::new(Type::ANY));
    }
    for name in BUILTIN_FUNCTION_NAMES {
        let obj = scope.lookup(name).unwrap();
        let obj_ref = obj.borrow_mut();
        assert!(obj_ref.ty.is_any());
    }
}

#[test]
fn test_resolve_program() {
    let mut program = parse_program("./src/resolver/test_data/assign.k").unwrap();
    let scope = resolve_program(&mut program);
    assert_eq!(scope.pkgpaths(), vec!["__main__".to_string()]);
    let main_scope = scope.main_scope().unwrap();
    let main_scope = main_scope.borrow_mut();
    assert!(main_scope.lookup("a").is_some());
    assert!(main_scope.lookup("b").is_some());
    assert!(main_scope.lookup("print").is_none());
}

#[test]
fn test_pkg_init_in_schema_resolve() {
    let mut program =
        load_program(&["./src/resolver/test_data/pkg_init_in_schema.k"], None).unwrap();
    let scope = resolve_program(&mut program);
    assert_eq!(
        scope.pkgpaths(),
        vec!["__main__".to_string(), "pkg".to_string()]
    );
    let module = &program.pkgs["pkg"][0];
    if let ast::Stmt::Schema(schema) = &module.body[1].node {
        if let ast::Stmt::SchemaAttr(attr) = &schema.body[0].node {
            let value = attr.value.as_ref().unwrap();
            if let ast::Expr::Schema(schema_expr) = &value.node {
                assert_eq!(schema_expr.name.node.names, vec!["Name".to_string()]);
            } else {
                panic!("test failed, expect schema expr, got {:?}", value)
            }
        } else {
            panic!(
                "test failed, expect schema attribute, got {:?}",
                schema.body[0]
            )
        }
    } else {
        panic!(
            "test failed, expect schema statement, got {:?}",
            module.body[1]
        )
    }
}

#[test]
fn test_resolve_program_fail() {
    let mut program = parse_program("./src/resolver/test_fail_data/config_expr.k").unwrap();
    let scope = resolve_program(&mut program);
    assert_eq!(scope.diagnostics.len(), 1);
    let diag = &scope.diagnostics[0];
    assert_eq!(diag.code, Some(DiagnosticId::Error(ErrorKind::TypeError)));
    assert_eq!(diag.messages.len(), 1);
    assert_eq!(diag.messages[0].message, "expect int, got {str:int(1)}");
}

#[test]
fn test_resolve_program_cycle_reference_fail() {
    let mut program = load_program(
        &["./src/resolver/test_fail_data/cycle_reference/file1.k"],
        None,
    )
    .unwrap();
    let scope = resolve_program(&mut program);
    let err_messages = [
        "There is a circular import reference between module file1 and file2",
        "There is a circular reference between schema SchemaBase and SchemaSub",
        "There is a circular reference between schema SchemaSub and SchemaBase",
        "There is a circular reference between rule RuleBase and RuleSub",
        "There is a circular reference between rule RuleSub and RuleBase",
    ];
    assert_eq!(scope.diagnostics.len(), err_messages.len());
    for (diag, msg) in scope.diagnostics.iter().zip(err_messages.iter()) {
        assert_eq!(diag.messages[0].message, msg.to_string(),);
    }
}

#[test]
fn test_lint() {
    let mut program = load_program(&["./src/resolver/test_data/import.k"], None).unwrap();
    pre_process_program(&mut program);
    let mut resolver = Resolver::new(
        &program,
        Options {
            raise_err: true,
            config_auto_fix: false,
            lint_check: true,
        },
    );
    resolver.resolve_import();
    resolver.check(kclvm_ast::MAIN_PKG);

    let root = &program.root.clone();
    let filename = root.clone() + "/import.k";

    let mut diagnostics: IndexSet<Diagnostic> = IndexSet::default();
    diagnostics.insert(Diagnostic {
        level: Level::Error,
        messages: vec![Message {
            pos: Position {
                filename: filename.clone(),
                line: 1,
                column: None,
            },
            style: Style::Line,
            message: format!(
                "Cannot find the module {} from {}",
                "abc",
                root.clone() + "/abc"
            ),
            note: None,
        }],
        code: Some(DiagnosticId::Error(ErrorKind::CannotFindModule)),
    });
    diagnostics.insert(Diagnostic {
        level: Level::Warning,
        messages: vec![Message {
            pos: Position {
                filename: filename.clone(),
                line: 17,
                column: None,
            },
            style: Style::Line,
            message: format!("Importstmt should be placed at the top of the module"),
            note: Some("Consider moving tihs statement to the top of the file".to_string()),
        }],
        code: Some(DiagnosticId::Warning(WarningKind::ImportPositionWarning)),
    });
    diagnostics.insert(Diagnostic {
        level: Level::Warning,
        messages: vec![Message {
            pos: Position {
                filename: filename.clone(),
                line: 3,
                column: None,
            },
            style: Style::Line,
            message: format!("Module '{}' is reimported multiple times.", "a",),
            note: Some("Consider removing this statement".to_string()),
        }],
        code: Some(DiagnosticId::Warning(WarningKind::ReimportWarning)),
    });
    diagnostics.insert(Diagnostic {
        level: Level::Warning,
        messages: vec![Message {
            pos: Position {
                filename: filename.clone(),
                line: 1,
                column: None,
            },
            style: Style::Line,
            message: format!("Module '{}' imported but unused.", "abc",),
            note: Some("Consider removing this statement".to_string()),
        }],
        code: Some(DiagnosticId::Warning(WarningKind::UnusedImportWarning)),
    });

    for (d1, d2) in resolver.handler.diagnostics.iter().zip(diagnostics.iter()) {
        assert_eq!(d1, d2);
    }
}
