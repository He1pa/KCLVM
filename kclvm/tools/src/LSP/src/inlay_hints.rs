use indexmap::IndexSet;
use kclvm_sema::core::global_state::GlobalState;
use kclvm_sema::core::symbol::{SymbolHint, SymbolHintKind};
use lsp_types::{
    InlayHint, InlayHintKind, InlayHintLabelPart, Position as LspPosition, Range, TextEdit,
};
use std::hash::Hash;

use crate::to_lsp::lsp_pos;

#[derive(Clone, Debug)]
struct KCLInlayHint {
    /// The position of this hint.
    pub position: LspPosition,

    /// An inlay hint label part allows for interactive and composite labels
    /// of inlay hints.
    pub part: InlayHintLabelPart,

    pub kind: InlayHintKind,

    /// Optional text edits that are performed when accepting(e.g. double-click in VSCode) this inlay hint.
    pub text_edits: Option<Vec<TextEdit>>,
}

impl Hash for KCLInlayHint {
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        self.position.line.hash(state);
        self.position.character.hash(state);
        self.part.value.hash(state);
    }
}

impl PartialEq for KCLInlayHint {
    fn eq(&self, other: &Self) -> bool {
        self.position == other.position && self.part.value == other.part.value
    }
}

impl Eq for KCLInlayHint {}

pub fn inlay_hints(file: &str, gs: &GlobalState) -> Option<Vec<InlayHint>> {
    let mut inlay_hints: IndexSet<KCLInlayHint> = Default::default();
    let sema_db = gs.get_sema_db();
    if let Some(file_sema) = sema_db.get_file_sema(file) {
        for hint in file_sema.get_hints() {
            inlay_hints.insert(generate_inlay_hint(hint));
        }
    }
    Some(
        inlay_hints
            .into_iter()
            .map(|h| into_lsp_inlay_hint(&h))
            .collect(),
    )
}

#[inline]
fn generate_inlay_hint(hint: &SymbolHint) -> KCLInlayHint {
    let (part, position, kind) = get_hint_label(hint);
    let text_edits = match hint.kind {
        SymbolHintKind::TypeHint(_) => Some(vec![TextEdit {
            range: Range {
                start: position.clone(),
                end: position.clone(),
            },
            new_text: part.value.clone(),
        }]),
        SymbolHintKind::VarHint(_) => None,
    };
    KCLInlayHint {
        position,
        part,
        kind,
        text_edits,
    }
}

#[inline]
fn into_lsp_inlay_hint(hint: &KCLInlayHint) -> InlayHint {
    InlayHint {
        position: hint.position.clone(),
        label: lsp_types::InlayHintLabel::LabelParts(vec![hint.part.clone()]),
        kind: Some(hint.kind),
        text_edits: hint.text_edits.clone(),
        tooltip: None,
        padding_left: None,
        padding_right: None,
        data: None,
    }
}

fn get_hint_label(hint: &SymbolHint) -> (InlayHintLabelPart, LspPosition, InlayHintKind) {
    match &hint.kind {
        SymbolHintKind::TypeHint(ty) => (
            InlayHintLabelPart {
                value: format!(": {ty}"),
                ..Default::default()
            },
            lsp_pos(&hint.pos),
            InlayHintKind::TYPE,
        ),
        SymbolHintKind::VarHint(var) => (
            InlayHintLabelPart {
                value: format!("{var}: "),
                ..Default::default()
            },
            lsp_pos(&hint.pos),
            InlayHintKind::PARAMETER,
        ),
    }
}

#[cfg(test)]
mod tests {
    use super::inlay_hints;
    use crate::tests::compile_test_file;

    #[macro_export]
    macro_rules! inlay_hints_test_snapshot {
        ($name:ident, $file:expr) => {
            #[test]
            fn $name() {
                let (file, _, _, gs) = compile_test_file($file);
                let res = inlay_hints(&file, &gs);
                insta::assert_snapshot!(format!("{:#?}", res));
            }
        };
    }

    inlay_hints_test_snapshot!(
        test_assign_stmt_type_hint,
        "src/test_data/inlay_hints/assign_stmt_type_hint/assign_stmt_type_hint.k"
    );

    inlay_hints_test_snapshot!(
        test_function_call_arg_hint,
        "src/test_data/inlay_hints/function_call/function_call.k"
    );

    inlay_hints_test_snapshot!(
        test_schema_arg_hint,
        "src/test_data/inlay_hints/schema_args/schema_args_hint.k"
    );

    // Temporary revert
    // inlay_hints_test_snapshot!(
    //     test_config_key_ty,
    //     "src/test_data/inlay_hints/config_key/config_key.k"
    // );
}
