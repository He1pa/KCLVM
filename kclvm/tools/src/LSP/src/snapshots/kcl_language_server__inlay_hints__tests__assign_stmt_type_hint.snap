---
source: tools/src/LSP/src/inlay_hints.rs
expression: "format!(\"{:#?}\", res)"
---
Some(
    [
        InlayHint {
            position: Position {
                line: 0,
                character: 1,
            },
            label: LabelParts(
                [
                    InlayHintLabelPart {
                        value: ": int",
                        tooltip: None,
                        location: None,
                        command: None,
                    },
                ],
            ),
            kind: Some(
                Type,
            ),
            text_edits: Some(
                [
                    TextEdit {
                        range: Range {
                            start: Position {
                                line: 0,
                                character: 1,
                            },
                            end: Position {
                                line: 0,
                                character: 1,
                            },
                        },
                        new_text: ": int",
                    },
                ],
            ),
            tooltip: None,
            padding_left: None,
            padding_right: None,
            data: None,
        },
        InlayHint {
            position: Position {
                line: 1,
                character: 1,
            },
            label: LabelParts(
                [
                    InlayHintLabelPart {
                        value: ": str",
                        tooltip: None,
                        location: None,
                        command: None,
                    },
                ],
            ),
            kind: Some(
                Type,
            ),
            text_edits: Some(
                [
                    TextEdit {
                        range: Range {
                            start: Position {
                                line: 1,
                                character: 1,
                            },
                            end: Position {
                                line: 1,
                                character: 1,
                            },
                        },
                        new_text: ": str",
                    },
                ],
            ),
            tooltip: None,
            padding_left: None,
            padding_right: None,
            data: None,
        },
        InlayHint {
            position: Position {
                line: 2,
                character: 1,
            },
            label: LabelParts(
                [
                    InlayHintLabelPart {
                        value: ": str",
                        tooltip: None,
                        location: None,
                        command: None,
                    },
                ],
            ),
            kind: Some(
                Type,
            ),
            text_edits: Some(
                [
                    TextEdit {
                        range: Range {
                            start: Position {
                                line: 2,
                                character: 1,
                            },
                            end: Position {
                                line: 2,
                                character: 1,
                            },
                        },
                        new_text: ": str",
                    },
                ],
            ),
            tooltip: None,
            padding_left: None,
            padding_right: None,
            data: None,
        },
        InlayHint {
            position: Position {
                line: 3,
                character: 1,
            },
            label: LabelParts(
                [
                    InlayHintLabelPart {
                        value: ": (int) -> int",
                        tooltip: None,
                        location: None,
                        command: None,
                    },
                ],
            ),
            kind: Some(
                Type,
            ),
            text_edits: Some(
                [
                    TextEdit {
                        range: Range {
                            start: Position {
                                line: 3,
                                character: 1,
                            },
                            end: Position {
                                line: 3,
                                character: 1,
                            },
                        },
                        new_text: ": (int) -> int",
                    },
                ],
            ),
            tooltip: None,
            padding_left: None,
            padding_right: None,
            data: None,
        },
        InlayHint {
            position: Position {
                line: 10,
                character: 1,
            },
            label: LabelParts(
                [
                    InlayHintLabelPart {
                        value: ": Name",
                        tooltip: None,
                        location: None,
                        command: None,
                    },
                ],
            ),
            kind: Some(
                Type,
            ),
            text_edits: Some(
                [
                    TextEdit {
                        range: Range {
                            start: Position {
                                line: 10,
                                character: 1,
                            },
                            end: Position {
                                line: 10,
                                character: 1,
                            },
                        },
                        new_text: ": Name",
                    },
                ],
            ),
            tooltip: None,
            padding_left: None,
            padding_right: None,
            data: None,
        },
        InlayHint {
            position: Position {
                line: 13,
                character: 2,
            },
            label: LabelParts(
                [
                    InlayHintLabelPart {
                        value: ": Name",
                        tooltip: None,
                        location: None,
                        command: None,
                    },
                ],
            ),
            kind: Some(
                Type,
            ),
            text_edits: Some(
                [
                    TextEdit {
                        range: Range {
                            start: Position {
                                line: 13,
                                character: 2,
                            },
                            end: Position {
                                line: 13,
                                character: 2,
                            },
                        },
                        new_text: ": Name",
                    },
                ],
            ),
            tooltip: None,
            padding_left: None,
            padding_right: None,
            data: None,
        },
        InlayHint {
            position: Position {
                line: 14,
                character: 4,
            },
            label: LabelParts(
                [
                    InlayHintLabelPart {
                        value: ": any",
                        tooltip: None,
                        location: None,
                        command: None,
                    },
                ],
            ),
            kind: Some(
                Type,
            ),
            text_edits: Some(
                [
                    TextEdit {
                        range: Range {
                            start: Position {
                                line: 14,
                                character: 4,
                            },
                            end: Position {
                                line: 14,
                                character: 4,
                            },
                        },
                        new_text: ": any",
                    },
                ],
            ),
            tooltip: None,
            padding_left: None,
            padding_right: None,
            data: None,
        },
        InlayHint {
            position: Position {
                line: 16,
                character: 1,
            },
            label: LabelParts(
                [
                    InlayHintLabelPart {
                        value: ": str",
                        tooltip: None,
                        location: None,
                        command: None,
                    },
                ],
            ),
            kind: Some(
                Type,
            ),
            text_edits: Some(
                [
                    TextEdit {
                        range: Range {
                            start: Position {
                                line: 16,
                                character: 1,
                            },
                            end: Position {
                                line: 16,
                                character: 1,
                            },
                        },
                        new_text: ": str",
                    },
                ],
            ),
            tooltip: None,
            padding_left: None,
            padding_right: None,
            data: None,
        },
        InlayHint {
            position: Position {
                line: 17,
                character: 2,
            },
            label: LabelParts(
                [
                    InlayHintLabelPart {
                        value: ": [int | str]",
                        tooltip: None,
                        location: None,
                        command: None,
                    },
                ],
            ),
            kind: Some(
                Type,
            ),
            text_edits: Some(
                [
                    TextEdit {
                        range: Range {
                            start: Position {
                                line: 17,
                                character: 2,
                            },
                            end: Position {
                                line: 17,
                                character: 2,
                            },
                        },
                        new_text: ": [int | str]",
                    },
                ],
            ),
            tooltip: None,
            padding_left: None,
            padding_right: None,
            data: None,
        },
        InlayHint {
            position: Position {
                line: 18,
                character: 2,
            },
            label: LabelParts(
                [
                    InlayHintLabelPart {
                        value: ": {str:str}",
                        tooltip: None,
                        location: None,
                        command: None,
                    },
                ],
            ),
            kind: Some(
                Type,
            ),
            text_edits: Some(
                [
                    TextEdit {
                        range: Range {
                            start: Position {
                                line: 18,
                                character: 2,
                            },
                            end: Position {
                                line: 18,
                                character: 2,
                            },
                        },
                        new_text: ": {str:str}",
                    },
                ],
            ),
            tooltip: None,
            padding_left: None,
            padding_right: None,
            data: None,
        },
        InlayHint {
            position: Position {
                line: 21,
                character: 3,
            },
            label: LabelParts(
                [
                    InlayHintLabelPart {
                        value: ": number_multiplier(1Ki)",
                        tooltip: None,
                        location: None,
                        command: None,
                    },
                ],
            ),
            kind: Some(
                Type,
            ),
            text_edits: Some(
                [
                    TextEdit {
                        range: Range {
                            start: Position {
                                line: 21,
                                character: 3,
                            },
                            end: Position {
                                line: 21,
                                character: 3,
                            },
                        },
                        new_text: ": number_multiplier(1Ki)",
                    },
                ],
            ),
            tooltip: None,
            padding_left: None,
            padding_right: None,
            data: None,
        },
        InlayHint {
            position: Position {
                line: 23,
                character: 4,
            },
            label: LabelParts(
                [
                    InlayHintLabelPart {
                        value: ": (any, any, any) -> any",
                        tooltip: None,
                        location: None,
                        command: None,
                    },
                ],
            ),
            kind: Some(
                Type,
            ),
            text_edits: Some(
                [
                    TextEdit {
                        range: Range {
                            start: Position {
                                line: 23,
                                character: 4,
                            },
                            end: Position {
                                line: 23,
                                character: 4,
                            },
                        },
                        new_text: ": (any, any, any) -> any",
                    },
                ],
            ),
            tooltip: None,
            padding_left: None,
            padding_right: None,
            data: None,
        },
    ],
)
